File: Step01_Basics/Lec6_Basic_Hashing/Program_1_map.cpp
#include <bits/stdc++.h>
using namespace std;

// Count elements of array using map

int main() {
    int n;
    cin >> n;
    int arr[n];
    for (int i = 0; i < n; i++) {
        cin >> arr[i];
    }

    // precompute
    map<int, int> mpp;
    for (int i = 0; i < n; i++) {
        mpp[arr[i]]++;
    }

    // interate in the map
    cout << "MAP STRUCTURE:" << endl;
    for (auto element : mpp) {
        cout << element.first << "->" << element.second << endl;
    }
    cout << endl;

    int q;
    cin >> q;
    while (q--) {
        int num;
        cin >> num;
        // fetch
        cout << num << " appears " << mpp[num] << " times" << endl;
    }
}

// every operation of map takes time-complexity: O(log n)

File: Step01_Basics/Lec6_Basic_Hashing/Program_1_map.java
package Step01_Basics.Lec6_Basic_Hashing;

import java.util.*;

// Count elements of array using map

public class Program_1_map {
    public static void main(String[] args) {
        int[] arr = {1, 2, 3, 1, 2, 3, 4, 1, 2, 3, 4, 5, 1, 2, 3, 4, 5, 6};
        int n = arr.length;

        // precompute
        SortedMap<Integer, Integer> map = new TreeMap<>();
        for (int element : arr) {
            map.put(element, map.getOrDefault(element, 0) + 1);
        }

        // iterate in the map
        System.out.println("Map Structure :");
        for (var mapElement : map.entrySet()) {
            System.out.println(mapElement.getKey() + " -> " + mapElement.getValue());
        }

        int[] query = {1, 4, 7};
        for (int num : query) {
            System.out.println(map.getOrDefault(num, 0));
        }
    }
}

// every operation of map takes time-complexity: O(log n)
